File: Python/backend/models/Log.py
from datetime import datetime
from .Cliente import Cliente
from .Utilizador import Utilizador
from .Marcacao import Marcacao
from .Pendente import Pendente

class Log:
    def __init__(self, data_hora=None, cliente=None, utilizador=None, marcacao=None, 
                 pendente=None, detalhes="", outro=None):
        """
        Construtor da classe Log
        Pode ser usado como:
        - Log() - construtor padrão
        - Log(data_hora, utilizador=utilizador, detalhes=detalhes) - apenas para utilizador
        - Log(data_hora, cliente=cliente, detalhes=detalhes) - apenas para cliente
        - Log(data_hora, marcacao=marcacao, detalhes=detalhes) - apenas para marcacao
        - Log(data_hora, pendente=pendente, detalhes=detalhes) - apenas para pendente
        - Log(outro=outro_log) - construtor de cópia
        """
        if outro is not None:
            # Construtor de cópia
            self.data_hora = outro.data_hora
            self.cliente = outro.cliente
            self.utilizador = outro.utilizador
            self.marcacao = outro.marcacao
            self.pendente = outro.pendente
            self.detalhes = outro.detalhes
        else:
            # Construtor padrão ou com parâmetros
            self.data_hora = data_hora if data_hora else datetime.now()
            self.cliente = cliente
            self.utilizador = utilizador
            self.marcacao = marcacao
            self.pendente = pendente
            self.detalhes = detalhes
            
            # Se marcacao foi fornecida, extrair cliente
            if marcacao is not None:
                self.cliente = marcacao.get_cliente()

    # Getters e Setters
    def get_data_hora(self):
        return self.data_hora

    def set_data_hora(self, data_hora):
        self.data_hora = data_hora

    def get_cliente(self):
        return self.cliente

    def set_cliente(self, cliente):
        self.cliente = cliente

    def get_utilizador(self):
        return self.utilizador

    def set_utilizador(self, utilizador):
        self.utilizador = utilizador

    def get_marcacao(self):
        return self.marcacao

    def set_marcacao(self, marcacao):
        self.marcacao = marcacao

    def get_pendente(self):
        return self.pendente

    def set_pendente(self, pendente):
        self.pendente = pendente

    def get_detalhes(self):
        return self.detalhes

    def set_detalhes(self, detalhes):
        self.detalhes = detalhes

    def clone(self):
        return Log(outro=self)

File: Java/src/models/Log.java
package models;

import java.time.LocalDateTime;

public class Log {
    private LocalDateTime dataHora;
    private Cliente cliente;
    private Utilizador utilizador;
    private Marcacao marcacao;
    private Pendente pendente;
    private String detalhes;

    // Construtores
    public Log() {
        this.dataHora = LocalDateTime.now();
        this.cliente = null;
        this.utilizador = null;
        this.marcacao = null;
        this.pendente = null;
        this.detalhes = "";
    }

    // apenas para utilizador
    public Log(LocalDateTime dataHora, Utilizador utilizador, String detalhes) {
        this.dataHora = dataHora;
        this.cliente = null;
        this.utilizador = utilizador;
        this.marcacao = null;
        this.pendente = null;
        this.detalhes = detalhes;
    }

    // apenas para cliente
    public Log(LocalDateTime dataHora, Cliente cliente, String detalhes) {
        this.dataHora = dataHora;
        this.cliente = cliente;
        this.utilizador = null;
        this.marcacao = null;
        this.pendente = null;
        this.detalhes = detalhes;
    }

    // apenas para marcacao
    public Log(LocalDateTime dataHora, Marcacao marcacao, String detalhes) {
        this.dataHora = dataHora;
        this.cliente = marcacao.getCliente();
        this.utilizador = null;
        this.marcacao = marcacao;
        this.pendente = null;
        this.detalhes = detalhes;
    }

    // apenas para pendente
    public Log(LocalDateTime dataHora, Pendente pendente, String detalhes) {
        this.dataHora = dataHora;
        this.cliente = null;
        this.utilizador = null;
        this.marcacao = null;
        this.pendente = pendente;
        this.detalhes = detalhes;
    }

    public Log(Log outro) {
        this.dataHora = outro.dataHora;
        this.cliente = outro.cliente;
        this.utilizador = outro.utilizador;
        this.marcacao = outro.marcacao;
        this.pendente = outro.pendente;
        this.detalhes = outro.detalhes;
    }

    // Getters e Setters
    public LocalDateTime getDataHora() {
        return dataHora;
    }

    public void setDataHora(LocalDateTime dataHora) {
        this.dataHora = dataHora;
    }

    public Cliente getCliente() {
        return cliente;
    }

    public void setCliente(Cliente cliente) {
        this.cliente = cliente;
    }

    public Utilizador getUtilizador() {
        return utilizador;
    }

    public void setUtilizador(Utilizador utilizador) {
        this.utilizador = utilizador;
    }

    public Marcacao getMarcacao() {
        return marcacao;
    }

    public void setMarcacao(Marcacao marcacao) {
        this.marcacao = marcacao;
    }

    public Pendente getPendente() {
        return pendente;
    }

    public void setPendente(Pendente pendente) {
        this.pendente = pendente;
    }

    public String getDetalhes() {
        return detalhes;
    }

    public void setDetalhes(String detalhes) {
        this.detalhes = detalhes;
    }

    @Override
    public Log clone() {
        return new Log(this);
    }

    @Override
    public String toString() {
        return "Log{" +
                "dataHora=" + dataHora +
                ", cliente='" + cliente + '\'' +
                ", utilizador='" + utilizador + '\'' +
                ", marcacao='" + marcacao + '\'' +
                ", detalhes='" + detalhes + '\'' +
                '}';
    }
}
